relayer/network_utils.go

```diff
@@ -0,0 +1,143 @@
+// or if the subnet supports all destinations, by the quora of all configured destinations.
+func InitializeConnectionsAndCheckStake(
+	logger logging.Logger,
+	network *peers.AppRequestNetwork,
+	cfg *config.Config,
+) error {
+	for _, sourceBlockchain := range cfg.SourceBlockchains {
+		if sourceBlockchain.GetSubnetID() == constants.PrimaryNetworkID {
+			if err := connectToPrimaryNetworkPeers(logger, network, cfg, sourceBlockchain); err != nil {
+				return fmt.Errorf(
+					"failed to connect to primary network peers: %w",
+					err,
+				)
+			}
+		} else {
+			if err := connectToNonPrimaryNetworkPeers(logger, network, cfg, sourceBlockchain); err != nil {
+				return fmt.Errorf(
+					"failed to connect to non-primary network peers: %w",
+					err,
+				)
+			}
+		}
+	}
+	return nil
+}
+
+// Connect to the validators of the source blockchain. For each destination blockchain,
+// verify that we have connected to a threshold of stake.
+func connectToNonPrimaryNetworkPeers(
+	logger logging.Logger,
+	network *peers.AppRequestNetwork,
+	cfg *config.Config,
+	sourceBlockchain *config.SourceBlockchain,
+) error {
+	subnetID := sourceBlockchain.GetSubnetID()
+	connectedValidators, err := network.ConnectToCanonicalValidators(subnetID)
+	if err != nil {
+		logger.Error(
+			"Failed to connect to canonical validators",
+			zap.String("subnetID", subnetID.String()),
+			zap.Error(err),
+		)
+		return err
+	}
+	for _, destination := range sourceBlockchain.SupportedDestinations {
+		blockchainID := destination.GetBlockchainID()
+		if ok, quorum, err := checkForSufficientConnectedStake(logger, cfg, connectedValidators, blockchainID); !ok {
+			logger.Error(
+				"Failed to connect to a threshold of stake",
+				zap.String("destinationBlockchainID", blockchainID.String()),
+				zap.Uint64("connectedWeight", connectedValidators.ConnectedWeight),
+				zap.Uint64("totalValidatorWeight", connectedValidators.TotalValidatorWeight),
+				zap.Any("warpQuorum", quorum),
+			)
+			return err
+		}
+	}
+	return nil
+}
+
+// Connect to the validators of the destination blockchains. Verify that we have connected
+// to a threshold of stake for each blockchain.
+func connectToPrimaryNetworkPeers(
+	logger logging.Logger,
+	network *peers.AppRequestNetwork,
+	cfg *config.Config,
+	sourceBlockchain *config.SourceBlockchain,
+) error {
+	for _, destination := range sourceBlockchain.SupportedDestinations {
+		blockchainID := destination.GetBlockchainID()
+		subnetID := cfg.GetSubnetID(blockchainID)
+		connectedValidators, err := network.ConnectToCanonicalValidators(subnetID)
+		if err != nil {
+			logger.Error(
+				"Failed to connect to canonical validators",
+				zap.String("subnetID", subnetID.String()),
+				zap.Error(err),
+			)
+			return err
+		}
+
+		if ok, quorum, err := checkForSufficientConnectedStake(logger, cfg, connectedValidators, blockchainID); !ok {
+			logger.Error(
```

Comment:
Thoughts on downgrading this log (and the above) to `Warn` since we're treating it as expected?

Reply:
Addressed! Thanks!